crypto/defaults.c

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright 1995-2023 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#include <stdio.h>
+#include "internal/thread_once.h"
+#include "internal/cryptlib.h"
+#include "internal/e_os.h"
+
+#if defined(_WIN32)
+
+/**
+ * @brief The directory where OpenSSL is installed.
+ */
+static char openssldir[MAX_PATH];
+
+/**
+ * @brief The directory where OpenSSL engines are located.
+ */
+
+static char enginesdir[MAX_PATH];
+
+/**
+ * @brief The directory where OpenSSL modules are located.
+ */
+static char modulesdir[MAX_PATH];
+
+/**
+ * @brief Get the list of Windows registry directories.
+ *
+ * This function retrieves a list of Windows registry directories.
+ *
+ * @return A pointer to a char array containing the registry directories.
+ */
+static char *get_windows_regdirs(char *dst, LPCWSTR valuename)
+{
+    DWORD keysize;
+    DWORD ktype;
+    HKEY hkey;
+    LSTATUS ret;
+    LPCWCH tempstr = NULL;
+    char *retval = NULL;
+
+    ret = RegOpenKeyEx(HKEY_LOCAL_MACHINE,
+                       TEXT("SOFTWARE\\Wow6432Node\\OpenSSL"), 0,
```

Comment:
Should this somehow include the version number in the key? i.e. so that you can have multiple instances of OpenSSL installed?

Reply:
I was trying to decide on that, we could certainly do that, but it creates another level of indirection, given that these can already be overriden by the OPENSSL_CONF, OPENSSL_MODULES, and OPENSSL_ENGINES environment variable.

Reply:
I think it is very likely that people will want multiple installations and to keep them separate. I also think the installer should be aware of what version it is installing. So I think it would be a really good idea to do this. I probably would not keep different *patch* versions distinct, but IMO different minor/major versions definitely should be.

Reply:
Suggested read order : 
 - HKEY_CURRENT_USER\SOFTWARE\\Wow6432Node\\OpenSSL\3.2
 - HKEY_LOCAL_MACHINE\\SOFTWARE\\Wow6432Node\\OpenSSL\3.2
 - HKEY_CURRENT_USER\SOFTWARE\\Wow6432Node\\OpenSSL
 - HKEY_LOCAL_MACHINE\SOFTWARE\\Wow6432Node\\OpenSSL

where : 
- `3.2` should be replaced by compiled version no.
- `Wow6432Node` should be removed on WIN64 define presence. 

Writing into HKEY_LOCAL_MACHINE implies "admin rights" for user and current running shell also.

Reply:
why the multi-tiered fallback?  Individual users can already override specific default values with OPENSSL_CONF / OPENSSL_ENGINES / OPENSSL_MODULES in the environment.  This seems like unneeded complication